src/main/java/net/snowflake/ingest/streaming/internal/ParquetRowBuffer.java

```diff
@@ -426,6 +399,101 @@ private float addRow(
+   * @param error insertion error object to populate if there is an error
+   * @param columnValueConsumer consumer to accept the column index and value
+   * @return the {@link ParseRowResult} containing the row size and the forked stats map
+   */
+  private ParseRowResult parseRow(
+      Map<String, Object> row,
+      long rowIndex,
+      InsertValidationResponse.InsertError error,
+      BiConsumer<Integer, Object> columnValueConsumer) {
+    // Create new empty stats just for the current row.
+    Map<String, RowBufferStats> forkedStatsMap = new HashMap<>();
+    statsMap.forEach((columnPath, stats) -> forkedStatsMap.put(columnPath, stats.forkEmpty()));
+    float size = 0F;
+    for (Map.Entry<String, Object> entry : row.entrySet()) {
+      String key = entry.getKey();
+      Object value = entry.getValue();
+      String columnName = LiteralQuoteUtils.unquoteColumnName(key);
+      ParquetColumn parquetColumn = fieldIndex.get(columnName);
+      int colIndex = parquetColumn.index;
+      ColumnMetadata column = parquetColumn.columnMetadata;
+      ParquetBufferValue valueWithSize =
+          (clientBufferParameters.isEnableIcebergStreaming()
+              ? IcebergParquetValueParser.parseColumnValueToParquet(
+                  value,
+                  parquetColumn.type,
+                  forkedStatsMap,
+                  subColumnFinder,
+                  defaultTimezone,
+                  rowIndex,
+                  error)
+              : SnowflakeParquetValueParser.parseColumnValueToParquet(
+                  value,
+                  column,
+                  parquetColumn.type.asPrimitiveType().getPrimitiveTypeName(),
+                  forkedStatsMap.get(columnName),
+                  defaultTimezone,
+                  rowIndex,
+                  clientBufferParameters.isEnableNewJsonParsingLogic()));
+      columnValueConsumer.accept(colIndex, valueWithSize.getValue());
+      size += valueWithSize.getSize();
+    }
+    return new ParseRowResult(size, forkedStatsMap);
+  }
+
+  @Override
+  public InsertValidationResponse validateRows(List<Map<String, Object>> rows) {
+
+    InsertValidationResponse response = new InsertValidationResponse();
+    int rowIndex = 0;
+    for (Map<String, Object> row : rows) {
+      InsertValidationResponse.InsertError error =
+          new InsertValidationResponse.InsertError(row, rowIndex);
+      try {
+        verifyInputColumns(row, error, rowIndex);
+        parseRow(row, rowIndex, error, (colIndex, value) -> {});
+      } catch (SFException e) {
+        error.setException(e);
+      } catch (Throwable e) {
+        error.setException(new SFException(e, ErrorCode.INTERNAL_ERROR, e.getMessage()));
+      }
+      if (isActualError(error)) {
+        response.addError(error);
+      }
+      rowIndex++;
+    }
+    return response;
+  }
+
+  private boolean isActualError(InsertValidationResponse.InsertError insertError) {
```

Comment:
Nit, should we rename this to `isDataError`? Let's make this clear this is specifically about data related errors and not server-side errors.